### DISCUSSION TITLE
use of opentelemetry Trace pipeline from agent and collector

### DISCUSSION THREAD

**Discussion by ajprakas** [open-telemetry/opentelemetry-java#2523] [Q&A]
I am using this example- https://github.com/open-telemetry/opentelemetry-collector/tree/master/examples/demo to instantiate a open telemetry agent and collector and publish spans to jaeger and zipkin backends via agent.How do I know which are the ports jaeger and zipkin receivers of agent are listening at so that I publish span to those port from my app.

after starting collector from docker-compose , I see agent, collector , jaeger and zipkin continers are running.

![image](https://user-images.githubusercontent.com/63052413/104553469-b3e65d00-5660-11eb-92ac-b199fa9b786b.png)

**Comment by anuraaga** [comment]
Hi @ajprakas - the ports for the receivers are defined in the collector config.yml files. For Zipkin, it explicitly sets 9411. For jaeger thrift http it is using the jaeger default port 14268 I believe

**Reply by ajprakas** [reply]
@anuraaga port 14268 its not working.
![image](https://user-images.githubusercontent.com/63052413/104684646-7b598880-571f-11eb-9ce7-d3a0e2790d7c.png)

 Below is my code for jaegerSpanexporter

JaegerGrpcSpanExporter jaegerGrpcSpanExporter = JaegerGrpcSpanExporter.builder().setServiceName(SERVICE_NAME)
                                                                .setChannel(ManagedChannelBuilder.forAddress("localhost", 14268).usePlaintext().build())
                                                                .build();
      this.openTelemetrySdk.getTracerManagement().addSpanProcessor(SimpleSpanProcessor.builder(jaegerGrpcSpanExporter).build());                                                          
      
below is agent receiver config
      
 receivers:
  opencensus:
  zipkin:
    endpoint: :9411
  jaeger:
    protocols:
      thrift_http:

**Reply by anuraaga** [reply]
@ajprakas Hmm - that demo configuration actually has a lot of legacy stuff in it I have filed https://github.com/open-telemetry/opentelemetry-collector/issues/2376 to update it.

In the meantime, I think the confusion is mainly that, because you are using the collector, definitely use our OTLP exporter from your app, not the jaeger exporter.

https://github.com/open-telemetry/opentelemetry-java/tree/master/exporters/otlp/trace

And configure the collector to use OTLP receiver (don't need any of the opencensus, zipkin, or jaeger receivers)

```
receivers:
  otlp:
    protocols:
      grpc:
```

**Reply by ajprakas** [reply]
@anuraaga  I am not clear how OLTP exporter works and how is it integrated in OTEL collectors? 
here is my otel collector receiver looks like

![image](https://user-images.githubusercontent.com/63052413/104876772-5d3d9380-597e-11eb-89cd-856ca2dbf050.png)


I think I am missing docker container for otlp receiver. What is the docker container to be used for oltp receiver. 
![image](https://user-images.githubusercontent.com/63052413/104876925-a5f54c80-597e-11eb-9683-a5bb683c7786.png)

Here is my instrumented code to use OTLP exporter

![image](https://user-images.githubusercontent.com/63052413/104876986-ce7d4680-597e-11eb-8909-cee92014027e.png)


Also Looks like OTLP  exporter is available for almost opentelemetry- api 0.9.1 . I was using 0.13.1 version of open-telemetry-api . Is there any chance OTLP  exporter  will be available for 0.13.1 version.

**Reply by anuraaga** [reply]
Hmm - yeah the config in the demo looks weird, not sure why it doesn't have the image name, but the definition is here

https://github.com/open-telemetry/opentelemetry-collector/blob/master/examples/demo/docker-compose.yaml#L37

The image is `otel/opentelemetry-collector` - https://hub.docker.com/r/otel/opentelemetry-collector. That image will run the collector based on the otel config.

> Also Looks like OTLP exporter is available for almost opentelemetry- api 0.9.1 . I was using 0.13.1 version of open-telemetry-api . Is there any chance OTLP exporter will be available for 0.13.1 version.

I think this is from before we did a rename of the artifacts. The current name of the artifact is `opentelemetry-exporter-otlp` (not `opentelemetry-exporters-otlp` with a "s")

**Reply by ajprakas** [reply]
@anuraaga I have tried your given example with version 0.13.1 . not sure What value I should provide as channel in OtlpGrpcSpanExporter. 
see my code below 
![image](https://user-images.githubusercontent.com/63052413/105056308-10e77600-5a9a-11eb-9522-f5a1f89fb081.png)

it says server not found

![image](https://user-images.githubusercontent.com/63052413/105056421-2b215400-5a9a-11eb-9842-c01c65bc4fa0.png)

Also I am using docker image - otel/opentelemetry-collector-dev:latest as per given example

do you have any working code as reference as given example is legacy.

**Reply by jkwatson** [reply]
We have a working example in the `examples/otlp` sub-project in this repository. It includes a docker-compose yaml file and a collector configuration that works.

**Reply by ajprakas** [reply]
Thanks @jkwatson given example helped me.
however service name is - OTLPRESOURCENOSERVICENAME in zipkin and nothing in jaeger as attcahed. 
How do we set service name in otlpSpanExporter?
![image](https://user-images.githubusercontent.com/63052413/105285362-40050100-5bda-11eb-99e7-5e5e3a90e888.png)

**Reply by ajprakas** [reply]
I was able to set service name via Resource and its working fine now. Thanks a ton 👍 

SdkTracerProvider.builder().setResource(Resource.create(Attributes.of(AttributeKey.stringKey("service.name"), "mongo-service"))).build()

**Reply by jkwatson** [reply]
Yes, the Resource is the correct way to set the service.name going forward. Glad you found it!
